pandablocks/ioc.py

```diff
@@ -0,0 +1,1720 @@
+
+            # Mask to remove every bit that isn't in the range we want
+            mask = (1 << bit_len) - 1
+
+            # Can't use proper numpy types, that's only available in 1.21+
+            val: Any = (packed[word_offset] >> bit_offset) & mask
+
+            if field_details.subtype == "int":
+                # First convert from 2's complement to offset, then add in offset.
+                # TODO: Test this with extreme values - int_max, int_min, etc.
+                val = np.int32((val ^ (1 << (bit_len - 1))) + (-1 << (bit_len - 1)))
+            else:
+                # Use shorter types, as these are used in waveform creation
+                if bit_len <= 8:
+                    val = np.uint8(val)
+                elif bit_len <= 16:
+                    val = np.uint16(val)
+
+            unpacked.append(val)
+
+        return unpacked
+
+    @staticmethod
+    def pack(
+        row_words: int,
+        table_records: Dict[str, _RecordInfo],
+        table_fields: Dict[str, TableFieldDetails],
+    ) -> List[str]:
+        """Pack the records based on the field definitions into the format PandA expects
+        for table writes.
+        TODO: parameter documentation
+
+        Returns:
+            List[str]: The list of data ready to be sent to PandA
+        """
+
+        packed = None
+
+        # Iterate over the zipped fields and their associated records to construct the
+        # packed array. Note that the MODE record is at the end of the table_records,
+        # and so will be ignored by zip() as it has no associated value in table_fields
+        for field_details, record_info in zip(
+            table_fields.values(), table_records.values()
+        ):
+            curr_val = record_info.record.get()
+            # PandA always handles tables in uint32 format
+            curr_val = np.uint32(curr_val)
+
+            if packed is None:
+                # Create 1-D array sufficiently long to exactly hold the entire table
+                packed = np.zeros((len(curr_val), row_words), dtype=np.uint32)
+            else:
+                # TODO: Probably shouldn't be an assert, use warning/error mechanism
+                assert len(packed) == len(curr_val), "Table waveform lengths mismatched"
+
+            offset = field_details.bit_low
+
+            # The word offset indicates which column this field is in
+            # (each column is one 32-bit word)
+            word_offset = offset // 32
+            # bit offset is location of field inside the word
+            bit_offset = offset & 0x1F
+
+            # Slice to get the column to apply the values to.
+            # bit shift the value to the relevant bits of the word
+            packed[:, word_offset] |= curr_val << bit_offset
+
+        # I'm surprised I need this assert here...
+        assert packed
+
+        # 2-D array -> 1-D array -> list[int] -> list[str]
+        return [str(x) for x in packed.flatten().tolist()]
```

Comment:
I ended up doing everything without base64. I just had a talk to Michael and he thinks that in most cases base64 will be more efficient for data transfer, but in some cases it will be less efficient, but probably a benefit overall for all but empty tables. I could add in the base64 requests again if wanted?

Reply:
No, lets leave it without base64, the sequencer tables are half empty at the moment so that should be more efficient